src/dict_header.F90

```diff
@@ -7,438 +7,528 @@ module dict_header
 ! pairs. This data structure is used to provide lookup features, e.g. cells and
 ! surfaces by name.
 !
-! The original version was roughly based on capabilities in the 'flibs' open
-! source package. However, it was rewritten from scratch so that it could be
-! used stand-alone without relying on the implementation of lists. As with
-! lists, it was considered writing a single dictionary used unlimited
-! polymorphism, but again compiler support is spotty and doesn't always prevent
-! duplication of code.
+! The implementation is based on Algorithm D from Knuth Vol. 3 Sec. 6.4 (open
+! addressing with double hashing). Hash tables sizes M are chosen such that M
+! and M - 2 are twin primes, which helps reduce clustering. An upper limit is
+! placed on the load factor to prevent exponential performance degradation as
+! the number of entries approaches the number of buckets.
 !===============================================================================
 
   implicit none
 
-  integer, parameter, private :: HASH_SIZE       = 4993
-  integer, parameter, private :: HASH_MULTIPLIER = 31
-  integer, parameter, private :: DICT_NULL       = -huge(0)
-  integer, parameter          :: DICT_KEY_LENGTH = 255
+  integer, parameter          :: EMPTY           = -huge(0)
+  integer, parameter, private :: DELETED         = -huge(0) + 1
+  integer, parameter, private :: KEY_CHAR_LENGTH = 255
+  integer, parameter, private :: TABLE_SIZES(30) = &
+       [5, 7, 13, 19, 43, 73, 151, 283, 571, 1153, 2269, 4519, 9013, 18043, &
+       36109, 72091, 144409, 288361, 576883, 1153459, 2307163, 4613893, &
+       9227641, 18455029, 36911011, 73819861, 147639589, 295279081, &
+       590559793, 1181116273]
```

Comment:
Did you come up with this sequence yourself? I looked at your Knuth reference but didn't see it there.

Reply:
I grabbed it from [another hash table implementation](https://github.com/anholt/hash_table/blob/master/hash_table.c) -- it seemed like a reasonable choice of twin primes that approximately doubled each time.

Reply:
Got it. Can you add a comment mentioning this in the above comment block? No need to include the license since you haven't used the code, per se, but good to list the source of these numbers for curious minds. Also would be good to mention that the growth is exponential with a factor of ~2x, so that building a dictionary is O(n) amortized.